pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -53,14 +56,19 @@ public class ConcurrentMapPartitionUpsertMetadataManager extends BasePartitionUp
   @VisibleForTesting
   final ConcurrentHashMap<Object, RecordLocation> _primaryKeyToRecordLocationMap = new ConcurrentHashMap<>();
 
+  // keep track of all segments that haven't reach stable state to persist snapshot
+  @VisibleForTesting
+  final PriorityBlockingQueue<SegmentInfo> _nonPersistedSegmentsQueue = new PriorityBlockingQueue<>();
```

Comment:
We can construct it only when TTL is enabled, and in the following places just check if it is `null`